File: results_helper.py
import logging
from urllib.parse import urljoin
import requests
from bs4 import BeautifulSoup

logger = logging.getLogger(__name__)

class ResultsChecking:
    BASE_URL = "http://125.16.54.154/mitsresults/resultug"
    ROMAN = {
        "I": "1",
        "II": "2",
        "III": "3",
        "IV": "4",
    }

    def __init__(self):
        self._cache = {}

    def _fetch_all(self):
        """Fetch and normalize all result entries from the results portal.
        Returns a list of tuples: (display_text, full_link, parts)
        where parts is the normalized split list of text elements.
        """
        html = requests.get(self.BASE_URL, timeout=20).text
        soup = BeautifulSoup(html, "lxml")
        wrapper = soup.find('div', class_='wrapper')
        if not wrapper:
            return []
        anchors = wrapper.find_all('a')
        entries = []
        for a in anchors:
            rel = a.get("href")
            name = a.get_text(strip=True)
            if not rel or not name:
                continue
            parts = [p.strip() for p in name.split('-')]
            # Expect like: [B.Tech, IV, II, R20, Regular, May, 2024]
            if len(parts) < 4:
                continue
            # Convert roman numerals for Year (idx 1) and Sem (idx 2) if present
            if parts[1].upper() in self.ROMAN:
                parts[1] = self.ROMAN[parts[1].upper()]
            if parts[2].upper() in self.ROMAN:
                parts[2] = self.ROMAN[parts[2].upper()]
            text_norm = "-".join(parts)
            full_link = urljoin(self.BASE_URL, rel)
            entries.append((text_norm, full_link, parts))
        return entries

    def get_results_link(self, collected):
        """Return a list of option display texts for given [reg, year, sem]."""
        if len(collected) < 3:
            return []
        reg, year, sem = collected[:3]
        key = (reg, year, sem)
        if key not in self._cache:
            all_entries = self._fetch_all()
            filtered = []
            for text, link, parts in all_entries:
                try:
                    if parts[1] == str(year) and parts[2] == str(sem) and reg in parts:
                        filtered.append((text, link))
                except Exception:
                    continue
            # Save filtered list
            self._cache[key] = filtered
        return [t for (t, _link) in self._cache[key]]

    def print_options(self, collected):
        """Given [reg, year, sem, option_index] return the final link URL."""
        if len(collected) < 4:
            return None
        reg, year, sem, idx = collected[:4]
        key = (reg, year, sem)
        if key not in self._cache:
            # Populate if not present
            self.get_results_link([reg, year, sem])
        items = self._cache.get(key, [])
        if not items:
            return None
        if not (0 <= int(idx) < len(items)):
            return None
        return items[int(idx)][1]

File: resultsHelper.js
import axios from 'axios'
import { load } from 'cheerio'

export class ResultsChecking {
  constructor() {
    this.BASE_URL = 'http://125.16.54.154/mitsresults/resultug'
    this.ROMAN = { I: '1', II: '2', III: '3', IV: '4' }
    this.cache = new Map()
  }

  async fetchAll() {
    const { data: html } = await axios.get(this.BASE_URL, { timeout: 20000 })
    const $ = load(html)
    const wrapper = $('div.wrapper')
    if (!wrapper || wrapper.length === 0) return []
    const anchors = wrapper.find('a')
    const entries = []
    anchors.each((i, el) => {
      const rel = $(el).attr('href')
      const name = $(el).text().trim()
      if (!rel || !name) return
      const parts = name.split('-').map(p => p.trim())
      if (parts.length < 4) return
      if (this.ROMAN[parts[1]?.toUpperCase()]) parts[1] = this.ROMAN[parts[1].toUpperCase()]
      if (this.ROMAN[parts[2]?.toUpperCase()]) parts[2] = this.ROMAN[parts[2].toUpperCase()]
      const textNorm = parts.join('-')
      const fullLink = new URL(rel, this.BASE_URL).toString()
      entries.push([textNorm, fullLink, parts])
    })
    return entries
  }

  getResultsLink(collected) {
    if (!collected || collected.length < 3) return []
    const [reg, year, sem] = collected
    const key = `${reg}-${year}-${sem}`
    if (!this.cache.has(key)) this.cache.set(key, [])
    if (this.cache.get(key).length === 0) {
      return this.fetchAll().then(all => {
        const filtered = []
        for (const e of all) {
          const parts = e[2]
          try {
            if (String(parts[1]) === String(year) && String(parts[2]) === String(sem) && parts.includes(reg)) filtered.push([e[0], e[1]])
          } catch {}
        }
        this.cache.set(key, filtered)
        return filtered.map(x => x[0])
      })
    }
    return this.cache.get(key).map(x => x[0])
  }

  printOptions(collected) {
    if (!collected || collected.length < 4) return null
    const [reg, year, sem, idx] = collected
    const key = `${reg}-${year}-${sem}`
    if (!this.cache.has(key) || this.cache.get(key).length === 0) return null
    const list = this.cache.get(key)
    const i = parseInt(idx, 10)
    if (!(i >= 0 && i < list.length)) return null
    return list[i][1]
  }
}